File: server/crates/persistence/src/book_copy/write_repo.rs
use anyhow::{Context, Result};
use chrono::Utc;
use domain::book_copy::{BookCopy, BookCopyId, BookCopyPrepared, BookCopyStatus};
use sqlx::PgExecutor;

use crate::book_copy::read_repo::BookCopyDbRow;

#[derive(sqlx::FromRow)]
pub struct BookCopyCreateResult {
    pub book_copy_id: i32,
}

pub(crate) async fn create_book_copy_with(
    executor: impl PgExecutor<'_>,
    insert: &BookCopyPrepared,
) -> Result<BookCopy> {
    let created_book_copy = sqlx::query_file_as!(
        BookCopyCreateResult,
        "sql/book_copy/commands/create.sql",
        insert.book_id.0,
        insert.status.to_string(),
        insert.barcode,
    )
    .fetch_one(executor)
    .await
    .context("Failed to create book copy")?;

    let now = Utc::now();
    let result = BookCopy {
        id: BookCopyId(created_book_copy.book_copy_id),
        barcode: insert.barcode.clone(),
        dt_created: now,
        dt_modified: now,
        book_id: insert.book_id,
        status: insert.status.clone(),
    };
    Ok(result)
}

pub(crate) async fn get_book_copy_by_barcode_for_update_with(
    executor: impl PgExecutor<'_>,
    barcode: &str,
) -> Result<Option<BookCopy>> {
    let row = sqlx::query_file_as!(
        BookCopyDbRow,
        "sql/book_copy/commands/get_by_barcode_for_update.sql",
        barcode
    )
    .fetch_optional(executor)
    .await
    .context("Failed to fetch book copy by barcode")?;

    row.map(BookCopy::try_from).transpose()
}

pub(crate) async fn update_book_copy_status_with(
    executor: impl PgExecutor<'_>,
    book_copy_id: BookCopyId,
    status: BookCopyStatus,
) -> Result<()> {
    sqlx::query_file!(
        "sql/book_copy/commands/update_status.sql",
        book_copy_id.0,
        status.to_string(),
    )
    .execute(executor)
    .await
    .context("Failed to update book copy status")?;

    Ok(())
}

File: servers/go/internal/persistence/postgres/bookcopy/write_repo.go
package bookcopy

import (
	"context"
	"fmt"
	"time"

	domain "github.com/christophercaldwell/model-architecture/go/internal/domain/bookcopy"
	domainbook "github.com/christophercaldwell/model-architecture/go/internal/domain/book"
	"github.com/jackc/pgx/v5"
)

type WriteRepo struct {
	tx pgx.Tx
}

func NewWriteRepo(tx pgx.Tx) *WriteRepo {
	return &WriteRepo{tx: tx}
}

func (r *WriteRepo) Create(ctx context.Context, prepared domain.BookCopyPrepared) (*domain.BookCopy, error) {
	const q = `
		INSERT INTO library.book_copy (book_id, status_id, barcode)
		VALUES (
			$1,
			(SELECT st.struct_type_id FROM library.struct_type st
			 WHERE st.group_name = 'book_copy_status' AND st.att_pub_ident = $2),
			$3
		)
		RETURNING book_copy_id`

	var bookCopyID int32
	err := r.tx.QueryRow(ctx, q, int32(prepared.BookID), string(prepared.Status), prepared.Barcode).Scan(&bookCopyID)
	if err != nil {
		return nil, fmt.Errorf("create book copy: %w", err)
	}
	now := time.Now()
	bc := domain.BookCopy{
		ID:         domain.BookCopyID(bookCopyID),
		Barcode:    prepared.Barcode,
		DtCreated:  now,
		DtModified: now,
		BookID:     domainbook.BookID(prepared.BookID),
		Status:     prepared.Status,
	}
	return &bc, nil
}

func (r *WriteRepo) GetByBarcodeForUpdate(ctx context.Context, barcode string) (*domain.BookCopy, error) {
	const q = `
		SELECT bc.book_copy_id, bc.barcode, bc.dt_created, bc.dt_modified, bc.book_id, st.att_pub_ident
		FROM library.book_copy bc
		JOIN library.struct_type st ON bc.status_id = st.struct_type_id
		WHERE bc.barcode = $1
		FOR UPDATE OF bc`

	var row bookCopyRow
	err := r.tx.QueryRow(ctx, q, barcode).Scan(
		&row.BookCopyID, &row.Barcode, &row.DtCreated, &row.DtModified, &row.BookID, &row.Status,
	)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("fetch book copy by barcode for update: %w", err)
	}
	return rowToDomain(row)
}

func (r *WriteRepo) UpdateStatus(ctx context.Context, id domain.BookCopyID, status domain.BookCopyStatus) error {
	const q = `
		UPDATE library.book_copy bc
		SET status_id = (
			SELECT st.struct_type_id FROM library.struct_type st
			WHERE st.group_name = 'book_copy_status' AND st.att_pub_ident = $2
		)
		WHERE bc.book_copy_id = $1`

	_, err := r.tx.Exec(ctx, q, int32(id), string(status))
	if err != nil {
		return fmt.Errorf("update book copy status: %w", err)
	}
	return nil
}
